diff --git MODULE.bazel MODULE.bazel
index edb49d0b..c16aa93d 100644
--- MODULE.bazel
+++ MODULE.bazel
@@ -43 +43,11 @@ bazel_dep(
 )
+
+bazel_dep(
+    name = "gloop",
+    repo_name = "do_not_use_for_gloop_visibility_only",
+)
+git_override(
+    module_name = "gloop",
+    remote = "https://github.com/abseil/gloop",
+    commit = "8eead4aaf1eb3685921b09e4fc03a69a31e35d53",
+)
diff --git absl/base/BUILD.bazel absl/base/BUILD.bazel
index 2e2630f2..0f57b452 100644
--- absl/base/BUILD.bazel
+++ absl/base/BUILD.bazel
@@ -59,2 +59,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -136,2 +137,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -161,2 +163,3 @@ cc_library(
         "//absl/base:__pkg__",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -191,2 +194,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -273,2 +277,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -453,2 +458,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -467,2 +473,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -549,2 +556,3 @@ cc_library(
         "//absl/base:__pkg__",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -584,2 +592,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -854,2 +863,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -953,2 +963,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -1014,2 +1025,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -1043,2 +1055,3 @@ cc_library(
         "//absl:__subpackages__",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
diff --git absl/container/BUILD.bazel absl/container/BUILD.bazel
index be10bb5d..60438d30 100644
--- absl/container/BUILD.bazel
+++ absl/container/BUILD.bazel
@@ -228,2 +228,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -445,2 +446,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -946,2 +948,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
@@ -964,2 +967,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
@@ -981,2 +985,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
@@ -998,2 +1003,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
@@ -1017,2 +1023,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
@@ -1033,2 +1040,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
@@ -1049,2 +1057,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
@@ -1066,2 +1075,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
@@ -1244,2 +1254,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
diff --git absl/debugging/BUILD.bazel absl/debugging/BUILD.bazel
index e6f11e84..5b7480b2 100644
--- absl/debugging/BUILD.bazel
+++ absl/debugging/BUILD.bazel
@@ -160,2 +160,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/log/internal:__pkg__",
@@ -223,3 +224,6 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
-    visibility = ["//absl:friends"],
+    visibility = [
+        "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
+    ],
     deps = [
@@ -241,2 +245,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
diff --git absl/flags/BUILD.bazel absl/flags/BUILD.bazel
index b5ecb7f7..c7163a6d 100644
--- absl/flags/BUILD.bazel
+++ absl/flags/BUILD.bazel
@@ -46,2 +46,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/flags:__pkg__",
@@ -66,2 +67,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/flags:__pkg__",
@@ -134,2 +136,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -173,2 +176,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/flags:__pkg__",
@@ -274,2 +278,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/flags:__pkg__",
diff --git absl/log/internal/BUILD.bazel absl/log/internal/BUILD.bazel
index ebacbc14..f8272537 100644
--- absl/log/internal/BUILD.bazel
+++ absl/log/internal/BUILD.bazel
@@ -29,2 +29,3 @@ package(
         ":internal_users",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -111,2 +112,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/log:__pkg__",
@@ -157,2 +159,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/log:__pkg__",
@@ -190,2 +193,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/log:__pkg__",
@@ -310,2 +314,3 @@ cc_library(
         ":structured_proto_users",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -384,2 +389,3 @@ cc_library(
         ":structured_proto_users",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -405,2 +411,3 @@ cc_library(
         ":structured_proto_users",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -468,2 +475,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/log:__subpackages__",
@@ -566,2 +574,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
diff --git absl/meta/BUILD.bazel absl/meta/BUILD.bazel
index 6e52c6ec..9aef4b2a 100644
--- absl/meta/BUILD.bazel
+++ absl/meta/BUILD.bazel
@@ -45,2 +45,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -71,2 +72,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
diff --git absl/profiling/BUILD.bazel absl/profiling/BUILD.bazel
index 34e3d254..b058c8f3 100644
--- absl/profiling/BUILD.bazel
+++ absl/profiling/BUILD.bazel
@@ -43,2 +43,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -81,2 +82,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -110,2 +112,3 @@ cc_library(
     visibility = [
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl:__subpackages__",
diff --git absl/random/internal/BUILD.bazel absl/random/internal/BUILD.bazel
index c2f2c367..0a2ee72c 100644
--- absl/random/internal/BUILD.bazel
+++ absl/random/internal/BUILD.bazel
@@ -32,2 +32,3 @@ default_package_visibility = [
     "//absl:friends",
+    "@do_not_use_for_gloop_visibility_only//:__subpackages__",
 ]
diff --git absl/strings/BUILD.bazel absl/strings/BUILD.bazel
index 6762ecda..15ce2e0f 100644
--- absl/strings/BUILD.bazel
+++ absl/strings/BUILD.bazel
@@ -433,3 +433,4 @@ cc_library(
         "//visibility:private",
-    ],
+       "@do_not_use_for_gloop_visibility_only//:__subpackages__",
+     ],
     deps = [
@@ -1591,2 +1592,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
diff --git absl/synchronization/BUILD.bazel absl/synchronization/BUILD.bazel
index 583eff2d..ffdf46e1 100644
--- absl/synchronization/BUILD.bazel
+++ absl/synchronization/BUILD.bazel
@@ -68,2 +68,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -299,2 +300,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
@@ -348,2 +350,3 @@ cc_library(
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
