diff --git MODULE.bazel MODULE.bazel
index cc60d224..e83fa2f6 100644
--- MODULE.bazel
+++ MODULE.bazel
@@ -40,3 +40,13 @@ bazel_dep(
     name = "googletest",
     version = "1.17.0.bcr.2",
 )
+
+bazel_dep(
+    name = "gloop",  // Dummy
+    repo_name = "do_not_use_for_gloop_visibility_only",
+)
+git_override(
+    module_name = "gloop",
+    remote = "https://github.com/abseil/gloop",
+    commit = "8eead4aaf1eb3685921b09e4fc03a69a31e35d53",
+)
diff --git absl/base/BUILD.bazel absl/base/BUILD.bazel
index 6ade1488..2f84caad 100644
--- absl/base/BUILD.bazel
+++ absl/base/BUILD.bazel
@@ -57,6 +57,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [":config"],
 )
@@ -131,6 +132,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":atomic_hook",
@@ -156,6 +158,7 @@ cc_library(
     visibility = [
         "//absl:friends",
         "//absl/base:__pkg__",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":base_internal",
@@ -186,6 +189,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":base_internal",
@@ -268,6 +272,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":config",
@@ -448,6 +453,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":config",
@@ -462,6 +468,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
 )
 
@@ -544,6 +551,7 @@ cc_library(
     visibility = [
         "//absl:friends",
         "//absl/base:__pkg__",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":base",
@@ -579,6 +587,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":base",
@@ -850,6 +859,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":config",
@@ -949,6 +959,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":config",
@@ -1010,6 +1021,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":config",
@@ -1040,6 +1052,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/base:config",
diff --git absl/container/BUILD.bazel absl/container/BUILD.bazel
index e748f04c..51c7951c 100644
--- absl/container/BUILD.bazel
+++ absl/container/BUILD.bazel
@@ -229,6 +229,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = ["//absl/types:compare"],
 )
@@ -446,6 +447,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":common",
@@ -947,6 +949,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
     ],
     deps = [
@@ -965,6 +968,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
     ],
     deps = [
@@ -982,6 +986,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
     ],
     deps = [
@@ -999,6 +1004,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
     ],
     deps = [
@@ -1018,6 +1024,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
     ],
     deps = [
@@ -1034,6 +1041,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
     ],
     deps = [
@@ -1050,6 +1058,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
     ],
     deps = [
@@ -1067,6 +1076,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
     ],
     deps = [
@@ -1246,6 +1256,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
     ],
     deps = [
diff --git absl/debugging/BUILD.bazel absl/debugging/BUILD.bazel
index e6f11e84..5b7480b2 100644
--- absl/debugging/BUILD.bazel
+++ absl/debugging/BUILD.bazel
@@ -158,6 +158,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/log/internal:__pkg__",
     ],
     deps = [
@@ -221,7 +222,10 @@ cc_library(
     ],
     copts = ABSL_DEFAULT_COPTS,
     linkopts = ABSL_DEFAULT_LINKOPTS,
-    visibility = ["//absl:friends"],
+    visibility = [
+        "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
+    ],
     deps = [
         "//absl/base:config",
         "//absl/base:core_headers",
@@ -239,6 +243,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
         "//absl/debugging:__pkg__",
     ],
diff --git absl/flags/BUILD.bazel absl/flags/BUILD.bazel
index b5ecb7f7..c7163a6d 100644
--- absl/flags/BUILD.bazel
+++ absl/flags/BUILD.bazel
@@ -44,6 +44,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/flags:__pkg__",
     ],
     deps = [
@@ -64,6 +65,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/flags:__pkg__",
         "//absl/log:__pkg__",
     ],
@@ -132,6 +134,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/base:config",
@@ -171,6 +174,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/flags:__pkg__",
         "//absl/flags/rust:__pkg__",
     ],
@@ -272,6 +276,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/flags:__pkg__",
     ],
     deps = [
diff --git absl/log/internal/BUILD.bazel absl/log/internal/BUILD.bazel
index 0344b56d..a787ab35 100644
--- absl/log/internal/BUILD.bazel
+++ absl/log/internal/BUILD.bazel
@@ -27,6 +27,7 @@ load(
 package(
     default_visibility = [
         ":internal_users",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     features = [
         "header_modules",
@@ -110,6 +111,7 @@ cc_library(
     visibility = [
         "//absl:friends",
         "//absl/log:__pkg__",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/base:config",
@@ -125,6 +127,7 @@ cc_library(
     visibility = [
         ":internal_users",
         "//absl/status:__pkg__",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/flags:flag",
@@ -160,6 +163,7 @@ cc_library(
     visibility = [
         "//absl:friends",
         "//absl/log:__pkg__",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/base:config",
@@ -193,6 +197,7 @@ cc_library(
     visibility = [
         "//absl:friends",
         "//absl/log:__pkg__",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":append_truncated",
@@ -312,6 +317,7 @@ cc_library(
     visibility = [
         ":internal_users",
         ":structured_proto_users",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":log_message",
@@ -386,6 +392,7 @@ cc_library(
     visibility = [
         ":internal_users",
         ":structured_proto_users",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":globals",
@@ -407,6 +414,7 @@ cc_library(
     visibility = [
         ":internal_users",
         ":structured_proto_users",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":test_helpers",
@@ -472,6 +480,7 @@ cc_library(
         "//absl:friends",
         "//absl/log:__subpackages__",
         "//absl/status:__pkg__",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":fnmatch",
@@ -569,6 +578,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/base:config",
diff --git absl/meta/BUILD.bazel absl/meta/BUILD.bazel
index 6e52c6ec..9aef4b2a 100644
--- absl/meta/BUILD.bazel
+++ absl/meta/BUILD.bazel
@@ -43,6 +43,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/base:config",
@@ -69,6 +70,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/base:config",
diff --git absl/profiling/BUILD.bazel absl/profiling/BUILD.bazel
index 34e3d254..b058c8f3 100644
--- absl/profiling/BUILD.bazel
+++ absl/profiling/BUILD.bazel
@@ -41,6 +41,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/base:config",
@@ -79,6 +80,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/base:config",
@@ -108,6 +110,7 @@ cc_library(
     copts = ABSL_DEFAULT_COPTS,
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl:__subpackages__",
         "//absl:friends",
     ],
diff --git absl/random/internal/BUILD.bazel absl/random/internal/BUILD.bazel
index 56377ed1..c229570a 100644
--- absl/random/internal/BUILD.bazel
+++ absl/random/internal/BUILD.bazel
@@ -30,6 +30,7 @@ load(
 default_package_visibility = [
     "//absl/random:__pkg__",
     "//absl:friends",
+    "@do_not_use_for_gloop_visibility_only//:__subpackages__",
 ]
 
 package(
diff --git absl/strings/BUILD.bazel absl/strings/BUILD.bazel
index 805cfc28..d0d3ee30 100644
--- absl/strings/BUILD.bazel
+++ absl/strings/BUILD.bazel
@@ -436,7 +436,8 @@ cc_library(
         "//absl:friends",
         "//absl/status:__pkg__",
         "//visibility:private",
-    ],
+       "@do_not_use_for_gloop_visibility_only//:__subpackages__",
+     ],
     deps = [
         ":string_view",
         "//absl/base:config",
@@ -1599,6 +1600,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":str_format",
diff --git absl/synchronization/BUILD.bazel absl/synchronization/BUILD.bazel
index 583eff2d..ffdf46e1 100644
--- absl/synchronization/BUILD.bazel
+++ absl/synchronization/BUILD.bazel
@@ -66,6 +66,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/base",
@@ -297,6 +298,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":synchronization",
@@ -346,6 +348,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":synchronization",
