diff --git MODULE.bazel MODULE.bazel
index cc60d224..e83fa2f6 100644
--- MODULE.bazel
+++ MODULE.bazel
@@ -42 +42,11 @@ bazel_dep(
 )
+
+bazel_dep(
+    name = "gloop",
+    repo_name = "do_not_use_for_gloop_visibility_only",
+)
+git_override(
+    module_name = "gloop",
+    remote = "https://github.com/abseil/gloop",
+    commit = "8eead4aaf1eb3685921b09e4fc03a69a31e35d53",
+)
diff --git absl/BUILD.bazel absl/BUILD.bazel
index e33d6489..7054fdee 100644
--- absl/BUILD.bazel
+++ absl/BUILD.bazel
@@ -66,3 +66,5 @@ package_group(
     name = "friends",
-    packages = [],
+    packages = [
+        "@do_not_use_for_gloop_visibility_only//...",
+    ],
 )
