diff --git MODULE.bazel MODULE.bazel
index cc60d224..e83fa2f6 100644
--- MODULE.bazel
+++ MODULE.bazel
@@ -40,3 +40,13 @@ bazel_dep(
     name = "googletest",
     version = "1.17.0.bcr.2",
 )
+
+bazel_dep(
+    name = "gloop",
+    repo_name = "do_not_use_for_gloop_visibility_only",
+)
+git_override(
+    module_name = "gloop",
+    remote = "https://github.com/abseil/gloop",
+    commit = "8eead4aaf1eb3685921b09e4fc03a69a31e35d53",
+)
diff --git absl/base/BUILD.bazel absl/base/BUILD.bazel
index 53bac42e..6dda4fff 100644
--- absl/base/BUILD.bazel
+++ absl/base/BUILD.bazel
@@ -57,6 +57,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [":config"],
 )
@@ -89,6 +90,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":config",
@@ -150,6 +152,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":atomic_hook",
@@ -175,6 +178,7 @@ cc_library(
     visibility = [
         "//absl:friends",
         "//absl/base:__pkg__",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":base_internal",
@@ -205,6 +209,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":base_internal",
@@ -287,6 +292,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":config",
@@ -467,6 +473,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":config",
@@ -481,6 +488,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
 )
 
@@ -563,6 +571,7 @@ cc_library(
     visibility = [
         "//absl:friends",
         "//absl/base:__pkg__",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":base",
@@ -598,6 +607,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":base",
@@ -869,6 +879,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":config",
@@ -968,6 +979,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":config",
@@ -1029,6 +1041,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":config",
@@ -1059,6 +1072,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/base:config",
diff --git absl/container/BUILD.bazel absl/container/BUILD.bazel
index a386a537..bd6df665 100644
--- absl/container/BUILD.bazel
+++ absl/container/BUILD.bazel
@@ -231,6 +231,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = ["//absl/types:compare"],
 )
@@ -448,6 +449,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":common",
@@ -949,6 +951,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
     ],
     deps = [
@@ -967,6 +970,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
     ],
     deps = [
@@ -984,6 +988,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
     ],
     deps = [
@@ -1001,6 +1006,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
     ],
     deps = [
@@ -1020,6 +1026,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
     ],
     deps = [
@@ -1036,6 +1043,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
     ],
     deps = [
@@ -1052,6 +1060,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
     ],
     deps = [
@@ -1069,6 +1078,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
     ],
     deps = [
@@ -1248,6 +1258,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
     ],
     deps = [
diff --git absl/debugging/BUILD.bazel absl/debugging/BUILD.bazel
index e6f11e84..5b7480b2 100644
--- absl/debugging/BUILD.bazel
+++ absl/debugging/BUILD.bazel
@@ -158,6 +158,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/log/internal:__pkg__",
     ],
     deps = [
@@ -221,7 +222,10 @@ cc_library(
     ],
     copts = ABSL_DEFAULT_COPTS,
     linkopts = ABSL_DEFAULT_LINKOPTS,
-    visibility = ["//absl:friends"],
+    visibility = [
+        "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
+    ],
     deps = [
         "//absl/base:config",
         "//absl/base:core_headers",
@@ -239,6 +243,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/container:__pkg__",
         "//absl/debugging:__pkg__",
     ],
diff --git absl/flags/BUILD.bazel absl/flags/BUILD.bazel
index b5ecb7f7..c7163a6d 100644
--- absl/flags/BUILD.bazel
+++ absl/flags/BUILD.bazel
@@ -44,6 +44,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/flags:__pkg__",
     ],
     deps = [
@@ -64,6 +65,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/flags:__pkg__",
         "//absl/log:__pkg__",
     ],
@@ -132,6 +134,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/base:config",
@@ -171,6 +174,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/flags:__pkg__",
         "//absl/flags/rust:__pkg__",
     ],
@@ -272,6 +276,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl/flags:__pkg__",
     ],
     deps = [
diff --git absl/log/internal/BUILD.bazel absl/log/internal/BUILD.bazel
index bc473308..53bce896 100644
--- absl/log/internal/BUILD.bazel
+++ absl/log/internal/BUILD.bazel
@@ -27,6 +27,7 @@ load(
 package(
     default_visibility = [
         ":internal_users",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     features = [
         "header_modules",
@@ -110,6 +111,7 @@ cc_library(
     visibility = [
         "//absl:friends",
         "//absl/log:__pkg__",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = ["//absl/base:config"],
 )
@@ -123,6 +125,7 @@ cc_library(
         ":internal_users",
         "//absl:friends",
         "//absl/status:__pkg__",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/flags:flag",
@@ -158,6 +161,7 @@ cc_library(
     visibility = [
         "//absl:friends",
         "//absl/log:__pkg__",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/base:config",
@@ -191,6 +195,7 @@ cc_library(
     visibility = [
         "//absl:friends",
         "//absl/log:__pkg__",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":append_truncated",
@@ -310,6 +315,7 @@ cc_library(
     visibility = [
         ":internal_users",
         ":structured_proto_users",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":log_message",
@@ -384,6 +390,7 @@ cc_library(
     visibility = [
         ":internal_users",
         ":structured_proto_users",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":globals",
@@ -405,6 +412,7 @@ cc_library(
     visibility = [
         ":internal_users",
         ":structured_proto_users",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":test_helpers",
@@ -470,6 +478,7 @@ cc_library(
         "//absl:friends",
         "//absl/log:__subpackages__",
         "//absl/status:__pkg__",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":fnmatch",
@@ -567,6 +576,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/base:config",
diff --git absl/meta/BUILD.bazel absl/meta/BUILD.bazel
index 6e52c6ec..9aef4b2a 100644
--- absl/meta/BUILD.bazel
+++ absl/meta/BUILD.bazel
@@ -43,6 +43,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/base:config",
@@ -69,6 +70,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/base:config",
diff --git absl/profiling/BUILD.bazel absl/profiling/BUILD.bazel
index 34e3d254..b058c8f3 100644
--- absl/profiling/BUILD.bazel
+++ absl/profiling/BUILD.bazel
@@ -41,6 +41,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/base:config",
@@ -79,6 +80,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/base:config",
@@ -108,6 +110,7 @@ cc_library(
     copts = ABSL_DEFAULT_COPTS,
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
         "//absl:__subpackages__",
         "//absl:friends",
     ],
diff --git absl/random/internal/BUILD.bazel absl/random/internal/BUILD.bazel
index 56377ed1..c229570a 100644
--- absl/random/internal/BUILD.bazel
+++ absl/random/internal/BUILD.bazel
@@ -30,6 +30,7 @@ load(
 default_package_visibility = [
     "//absl/random:__pkg__",
     "//absl:friends",
+    "@do_not_use_for_gloop_visibility_only//:__subpackages__",
 ]
 
 package(
diff --git absl/strings/BUILD.bazel absl/strings/BUILD.bazel
index def78d53..70449873 100644
--- absl/strings/BUILD.bazel
+++ absl/strings/BUILD.bazel
@@ -435,6 +435,7 @@ cc_library(
     visibility = [
         "//absl:friends",
         "//absl/status:__pkg__",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":string_view",
@@ -1599,6 +1600,7 @@ cc_library(
     visibility = [
         "//absl:__subpackages__",
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":str_format",
diff --git absl/synchronization/BUILD.bazel absl/synchronization/BUILD.bazel
index 583eff2d..ffdf46e1 100644
--- absl/synchronization/BUILD.bazel
+++ absl/synchronization/BUILD.bazel
@@ -66,6 +66,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         "//absl/base",
@@ -297,6 +298,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":synchronization",
@@ -346,6 +348,7 @@ cc_library(
     linkopts = ABSL_DEFAULT_LINKOPTS,
     visibility = [
         "//absl:friends",
+        "@do_not_use_for_gloop_visibility_only//:__subpackages__",
     ],
     deps = [
         ":synchronization",
